XamarinCommunityToolkit/Effects/Touch/GestureManager.shared.cs

```diff
@@ -0,0 +1,548 @@
+				? nameof(TouchState.Pressed)
+				: hoverState == HoverState.Hovered
+					? nameof(HoverState.Hovered)
+					: nameof(TouchState.Normal);
+
+			VisualStateManager.GoToState(visualElement, state);
+		}
+
+		async Task SetBackgroundImageAsync(TouchEffect sender, TouchState touchState, HoverState hoverState, int duration, CancellationToken token)
+		{
+			var normalBackgroundImageSource = sender.NormalBackgroundImageSource;
+			var pressedBackgroundImageSource = sender.PressedBackgroundImageSource;
+			var hoveredBackgroundImageSource = sender.HoveredBackgroundImageSource;
+
+			if (normalBackgroundImageSource == null &&
+				pressedBackgroundImageSource == null &&
+				hoveredBackgroundImageSource == null)
+				return;
+
+			var aspect = sender.NormalBackgroundImageAspect;
+			var source = normalBackgroundImageSource;
+			if (touchState == TouchState.Pressed)
+			{
+				aspect = sender.PressedBackgroundImageAspect;
+				source = pressedBackgroundImageSource;
+			}
+			else if (hoverState == HoverState.Hovered)
+			{
+				if (sender.Control.IsSet(TouchEffect.HoveredBackgroundImageAspectProperty))
+					aspect = sender.HoveredBackgroundImageAspect;
+
+				if (sender.Control.IsSet(TouchEffect.HoveredBackgroundImageSourceProperty))
+					source = hoveredBackgroundImageSource;
+			}
+
+			try
+			{
+				if (sender.ShouldSetImageOnAnimationEnd)
+					await Task.Delay(duration, token);
+			}
+			catch (TaskCanceledException)
+			{
+				return;
+			}
+
+			if (sender.Control is Image image)
+			{
+				using (image.Batch())
+				{
+					image.Aspect = aspect;
+					image.Source = source;
+				}
+			}
+		}
+
+		async Task SetBackgroundColorAsync(TouchEffect sender, TouchState touchState, HoverState hoverState, int duration, Easing easing)
+		{
+			var normalBackgroundColor = sender.NormalBackgroundColor;
+			var pressedBackgroundColor = sender.PressedBackgroundColor;
+			var hoveredBackgroundColor = sender.HoveredBackgroundColor;
+
+			if (normalBackgroundColor == Color.Default &&
+				pressedBackgroundColor == Color.Default &&
+				hoveredBackgroundColor == Color.Default)
+				return;
+
+			var control = sender.Control;
+			if (defaultBackgroundColor == default(Color))
+				defaultBackgroundColor = control.BackgroundColor;
+
+			var color = GetBackgroundColor(normalBackgroundColor);
+
+			if (touchState == TouchState.Pressed)
+				color = GetBackgroundColor(pressedBackgroundColor);
+			else if (hoverState == HoverState.Hovered && sender.Control.IsSet(TouchEffect.HoveredBackgroundColorProperty))
+				color = GetBackgroundColor(hoveredBackgroundColor);
+
+			if (duration <= 0)
+			{
+				control.BackgroundColor = color;
+				return;
+			}
+
+			var animationCompletionSource = new TaskCompletionSource<bool>();
+			new Animation
+            {
+				{ 0, 1,  new Animation(v => control.BackgroundColor = new Color(v, control.BackgroundColor.G, control.BackgroundColor.B, control.BackgroundColor.A), control.BackgroundColor.R, color.R) },
```

Comment:
to make it better, I think that we could move the `control.BackgroundColor` or each `control.BackgroundColor.X` into a local variable. What do you think?

Reply:
Unfortunately, we cannot do it because 4 parallel animations change different color parameters R G B A.
We should use the relevant value directly from the control's background color